src/infrastructure/channels/reliable_channel.rs

```diff
@@ -0,0 +1,146 @@
+            error!(
+                "Packet too large: Attempting to send {}, max={}",
+                payload.len(),
+                self.config.max_packet_size
+            );
+            Err(PacketErrorKind::ExceededMaxPacketSize)?;
+        }
+
+        // queue congestion data.
+        self.congestion_data.insert(
+            CongestionData::new(self.seq_num, Instant::now()),
+            self.seq_num,
+        );
+
+        // queue packet for awaiting acknowledgement.
+        self.waiting_packets.enqueue(self.seq_num, &payload);
+
+        // calculate size for or packet data.
+        let payload_length = payload.len() as u16; /* safe cast because max packet size is u16 */
+        let packet_data_size: u16 = Fragmentation::total_fragments_needed(payload_length, self.config.fragment_size);
+        let mut packet_data = PacketData::with_capacity(packet_data_size as usize);
+
+        let packet_type = if packet_data_size > 1 {
+            PacketTypeId::Fragment
+        }else {
+            PacketTypeId::Packet
+        };
+
+        // create our reliable header and write it to an buffer.
+        let header = AckedPacketHeader::new(StandardHeader::new(delivery_method, packet_type), self.seq_num, self.their_acks.last_seq, self.their_acks.field);
+        let mut buffer = Vec::with_capacity(header.size() as usize);
+        header.parse(&mut buffer)?;
+
+        // spit the packet if the payload length is greater than the allowed fragment size.
+        if payload_length <= self.config.fragment_size {
+            packet_data.add_fragment(&buffer, payload)?;
+        } else {
+            Fragmentation::spit_into_fragments(payload, header, &mut packet_data, &self.config)?;
+        }
+
+        // increase local sequence number.
+        self.seq_num = self.seq_num.wrapping_add(1);
+
+        Ok(packet_data)
+    }
+
+    /// Process a packet on receive.
+    ///
+    /// 1. Read reliable header.
+    /// 2. Update acknowledgement data.
+    /// 3. Calculate RTT time.
+    /// 4. Update dropped packets.
+    fn process_packet(&mut self, cursor: &mut Cursor<&[u8]>) -> NetworkResult<Vec<u8>> {
+        cursor.set_position(0);
+        let acked_header = AckedPacketHeader::read(cursor)?;
+
+        // get the slice of payload from the received data.
+        let mut payload = Vec::new();
+        cursor.read_to_end(&mut payload)?;
+
+        self.their_acks.ack(acked_header.seq);
+
+        // update congestion information.
+        let congestion_data = self.congestion_data.get_mut(acked_header.ack_seq());
+        self.network_quality_measurer.get_rtt(congestion_data);
+
+        // Update dropped packets if there are any.
+        let dropped_packets = self
+            .waiting_packets
+            .ack(acked_header.ack_seq(), acked_header.ack_field());
+
+        self.dropped_packets = dropped_packets.into_iter().map(|(_, p)| p).collect();
+        // TODO: resent packets if there are dropped packets
```

Comment:
Can we make an issue for this?

Reply:
Jup, needs to be done though, before release, I have those todo's for me so that I know I need to do some work on that still.